### DISCUSSION TITLE
useFetch: loading extra data (pagination) on top of the existing data

### DISCUSSION THREAD

**Discussion by RobertBReich** [neuefische/questions#139] [Web]
`

import axios from 'axios';
import {useEffect, useState} from 'react';

export default function useFetch(url) {
	const [state, setState] = useState({
		data: null,
		loading: false,
		error: null,
	});

	useEffect(() => {
		setState(({data}) => ({
			data,
			error: null,
			loading: true,
		}));
		axios
			.get(url)
			.then(({data}) => {
				setState({
					error: null,
					data: data,
					loading: false,
				});
			})
			.catch(error => {
				setState(({data}) => ({
					data,
					error,
					loading: false,
				}));
			});
	}, [url]);

	return state;
}
`

So basically the useFetch works like a charm for everything else. But now i want to load the 2nd page of data on click (pageCounter + 1) and i'm standing "wie der Ochs vorm Walde". 😂

A good tip anyone?

`
export default function Movies() {
	const API_KEY = process.env.API_KEY;
	let pageCounter = 1;
	let GET_MOVIES = "https://api.themoviedb.org/3/discover/movie?api_key=${API_KEY}&language=en-US&sort_by=popularity.desc&include_adult=false&include_video=false&page=${pageCounter}";

	const {loading, error, data} = useFetch(GET_MOVIES);

        loadMoreContent() {}

	return (
		<Layout>
			<Head>
				<title key="title">Movie Bookmark App</title>
				<meta
					key="description"
					name="description"
					content="Robert Reichs capstone project"
				/>
			</Head>
			<HaEins>Movies</HaEins>
			{loading && <p>Loading...</p>}
			{error && <p>The content could not be loaded. Please try again.</p>}
			{data && <MovieList data={data.results} />}
			<button onClick={loadMoreContent}>load more content</button> 
		</Layout>
	);
}
`

**Comment by JessicaLoers** [comment]
**Hi @RobertBReich** 
 
I haven't  the ONE solution for your project but an example from the session **useEffect with fetch** that will lead you on the right path.
In your project you use `zustand` and `axios`- so a little "Transferleistung" to the solution is necessary 😉

Here is the [Link zu the Codesandbox ](https://codesandbox.io/s/useeffect-with-fetch-o27qj5?file=/src/App.js)- have fun.

**Comment by alexclaes** [comment]
Hi @RobertBReich,

you didn't take time to formulate your question. Therefore your goal is not clear and it's very hard to read & understand your code.

The example provided by @JessicaLoers  is definitiv the right approach. This logic will replace all shown items with a completely new list. If you wish to extend an existing list, you would need to adapt the logic in `useFetch`.

It's not easy to read your example code, but it seems you are trying to call `loadMoreContent()`  right before the `return` statement. This will break the app. You need to wrap such function calls into `useEffect`. 

https://reactjs.org/docs/hooks-effect.html

**Comment by RobertBReich** [comment]
Not very helpful. 

It's not a useEffect question, it's about useFetch.js that is provided in our template. Alex said, that a coach has written it, so there is no online documentation for it. This -> [https://use-http.com/#/](https://use-http.com/#/) looks similar, but i think its not the same.

I tried a custom fetch() inside my loadMoreContent function and i get the data loaded - but the view doesn't update itself.

What I need is to call the useFetch again, with a new URL and an option to expand the data i already have. So at the start i have 20 entries, after that i want 40, 60 and so on.

And yes, the code in the question wasn't formatted very good. For some reason the first lines were plain text... Sorry for that.

**Comment by alexclaes** [comment]
Oh, both answer are indeed very helpful! 

They point you towards `useEffect`. You shouldn't be so sure, that this is not a `useEffect` question.

You described the required logic yourself:

> What I need is to call the useFetch again, with a new URL and an option to expand the data i already have. So at the start i have 20 entries, after that i want 40, 60 and so on.

You also described the problem precisely

>  but the view doesn't update itself.

Depending on your approach you might not even need to change anything in `useFetch`, like with this rough implementation: 

https://codesandbox.io/s/robert-useeffect-usefetch-j49j3l?file=/src/App.js
